ENH: ILP64 BLAS/LAPACK support

ev-br opened the pull request:
#### Reference issue
<!--Example: Closes gh-WXYZ.-->

towards https://github.com/scipy/scipy/issues/21889

fixes https://github.com/scipy/scipy/issues/23219

#### What does this implement/fix?
<!--Please explain your changes.-->

Add changes needed to build SciPy with 64-bit (ILP64) BLAS and LAPACK.

The main bulk of work for LAPACK detection is done by Ralf, see the commit history. (I squashed several commits to simplify rebasing, please let me know if some useful history got lost @rgommers).

On top of Ralf's work:
- hammer the MKL ILP64 detection into shape
- add source changes needed for the with-ILP64 MKL test run run to completion.
- EDIT: `scipy-openblas64` now works, too.

Am marking this PR as an enhancement even though we had all of this before moving off distutils to meson, so this PR just restores the status quo ante :-).

#### Additional information
<!--Any additional information you think is important.-->

The approach here is the option 2 of gh-21889 : we expect a LAPACK provider to provide two sets of symbols, 32-bit and 64-bit. The former must be always available, the latter are requested via a meson option.

The approach does not work for a general OpenBLAS build at the moment, so this PR only enables it for from-source builds with MKL and `scipy-openblas64`. See the `linux_blas_ilp64.yml` workflow for details.

How to test this: sadly, we cannot easily test linear algebra with sizes $> 2^{31}$ on CI. At the moment the only CI indicator is that the test run does not segfault. Locally, there's a test in linalg, `test_basic/test_large_vector`, which fails without ILP64 and passes with it.

Potential follow-ups:

- fix up "slow" tests. ATM `dev.py test -m "full"` fails with ILP64 for trivial reasons
- currently several submodules (optimize, special, cluster, odr) are using 32-bit integer size linear algebra; it is not very clear which of these really need 64-bit integers (SLSQP with $2^{31}$ variables, cough.
- generalize to other lapacks: Accelerate et al
- resolve https://github.com/scipy/scipy/issues/20002, provide a "canonical" header for LAPACK symbols to use from C/C++

Bigger issues:

- Decide on `cython_blas`  and  `cython_lapack`. Do we provide additional `cython_blas64` and `cython_lapack64` cython extensions?
- Do something in/with openblas which is not `scipy-openblas{32,64}`, so that we can have ILP64 by default, maybe even in the official wheels. That needs coordination with OpenBLAS and NumPy, it seems.

rgommers committed:
CI: add CI jobs for different BLAS configs

Adapted largely from NumPy's `linux_blas.yml`

rgommers committed:
DEBUG: run BLAS jobs on fork

rgommers committed:
Squash several commits to simplify rebasing

ci fix

Avoid pytest-cov, and install python311 on OpenSUSE

CI: enable Debian libblas/liblapack job

CI: add Fedora jobs for OpenBLAS and FlexiBLAS

OpenBLAS is detected without pkg-config, because Fedora does not
ship an openblas.pc file. FlexiBLAS is detected through pkg-config.

CI: add job using BLIS

CI: add MKL job

CI: add Accelerate (NEWLAPACK LP64) job

ENH: enable new macOS Accelerate BLAS/LAPACK support

BUG: fix issue with picking up use-g77-abi flag

rgommers committed:
WIP: ENH: enable ILP64 BLAS/LAPACK support again

This was left out initially when adding support for building with
Meson. The numpy.distutils based build supported ILP64 builds with
OpenBLAS and its `64_` symbol suffix scheme. We now add this back
and make it more generic, so that it also supports the `_64` scheme
of MKL as well as the `$NEWLAPACK_ILP64` scheme from Accelerate.

rgommers committed:
Squash several commits to simplify rebasing

DEBUG: enable macos job

BUG: ensure F77 ABI wrappers deal with BLAS symbol suffix correctly

This avoids build errors that look like:
```
../scipy/_build_utils/src/wrap_g77_abi_c.c:13:16: error: use of undeclared identifier 'cblas_cdotc_sub'
```
when one uses Accelerate with the `$NEWLAPACK` symbol suffix for example.

[ci skip]

WIP: ENH: more symbol renaming needed for BLAS symbol suffix support

rgommers committed:
ENH: use Meson's BLAS and LAPACK support

ev-br committed:
BLD: interpolate: build FITPACK with 32- or 64-bit ints

ev-br committed:
BUG: spatial: Qhull does not need linking to LAPACK

`qhull.pyx` uses cython_lapack, so linking is done at the build time.
NB: and if anything, Qhull might not be 64-bit int ready, so it'll
probably need an LP64 LAPACK.

ev-br committed:
TST: linalg: bump test tolerances

Violations observed on linux with MKL in the ILP64 mode.

ev-br committed:
ENH: integrate: make lsoda.pyf and vode.pyf int64-aware

ev-br committed:
BLD: integrate: add int64 flags to ODE wrappers (lsoda, dop, vode, zvode)

ev-br committed:
BLD: sparse/arpack: add an int64 flag

ev-br committed:
BLD: special: link to LP64-bit LAPACK, use cython_special in ellip_harm

ev-br committed:
MAINT: special: no need to link to LAPACK explicitly with cython_lapack

ev-br committed:
BLD: fix up MKL ILP64 detection

ev-br committed:
BLD: linalg, sparse/linalg: link with g77_abi_wrappers

ev-br committed:
BLD: optimize: link lbfgsb to 32-bit LP64 LAPACK

It otherwise segfaults. Needs using BLAS_FUNC maybe?

ev-br committed:
MAINT: interpolate: link the __fitpack C extension to 64-bit LAPACK

rgommers committed:
CI: separate the MKL ILP64 job

ev-br committed:
CI: tweak the ILP64 job

ev-br committed:
CI: temp remove linux_blas.yml workflow

ev-br committed:
BLD: fix a small error in meson.build

ev-br committed:
Revert "MAINT: special: no need to link to LAPACK explicitly with cython_lapack"

This reverts commit 8b60d82fee4dc06e3acd500e81bd8eb9bb220a8f.

ev-br committed:
Revert "BLD: special: link to LP64-bit LAPACK, use cython_special in ellip_harm"

This reverts commit 043a40e5787b146260ece50f5692ba9dcd5c6e4d.

ev-br committed:
BLD: special: link explicitly to 32-bit LAPACK

rgommers commented:
Nice to see progress on this, thanks @ev-br!

Once CI is happy, I think it makes sense to split off the changes to `.c|h|pyf` files and merge those, since that can immediately be done. 

> Am marking this PR as an enhancement even though we had all of this before moving off distutils to meson, so this PR just restores the status quo ante :-).

Not really, I am fairly sure that things only ever worked with a pretty janky OpenBLAS setup, not with MKL or Accelerate (different symbol name conventions).

ev-br committed:
WIP/BLD: try restoring the "regular" builds

ev-br committed:
WIP: revert ILP64 related changes in integrate, odr and arpack

ev-br committed:
BLD: force name mangling in ILP64 builds of BLAS and LAPACK

32-bit LAPACK builds generate the `blas_lapack_wrapper.so` library
which contains trampolines from F_FUNC to BLAS_FUNC mangling
(i.e. create dgemm_ as a synonym for scipy_dgemm_ etc).
Then f2py uses F_FUNC mangled symbols, and which get picked up
from the wrappers.

With ILP64 builds this approach may lead to symbol clashes:
at least the MKL _ilp64.so variants contain two kinds of symbols:
both `dgemm_64_` and `dgemm_`.
We want to only use the former (is dgemm_ 32- or 64-bit?).

Therefore, we instead force the BLAS_FUNC mangling at the f2py level:
inject a 'usercode' section into f2py-generated C files and replace
the F_FUNC define with BLAS_FUNC.
This relies on the fact that f2py includes the usercode section _after_
its F_FUNC define.

ev-br committed:
MAINT: interpolate: propagate C flags for ILP64

ev-br committed:
MAINT: slightly minimize the diff size

ev-br committed:
CI: remove the SDL MKL LP64 build for now

ev-br committed:
BLD: remove blas-order, simplify lapacl detection

ev-br committed:
CI: adjust mkl job for the lack of -Dblas-order

ev-br committed:
BUG: _build_utils: new Accelerate does not seem to need NO_APPEND_FORTRAN

ev-br committed:
MAINT: remove unused code

ev-br commented:
Okay, some progress: CI is all green for both existing jobs and the added ILP64 job;
Locally, the smoking gun test is passing, too:

   $ python dev.py build -C-Dblas=mkl -C-Duse-ilp64=true
   $ python dev.py test -s linalg -- -k'large_vector'

This test is skipped if ILP64 BLAS is not available, and fails if it is available but
LP64 routines get used instead.

To get there,

- removed the blas-order selection;
- reverted the ILP64 usage in most submodules apart from linalg
- simplified the ILP64 linking and made it less reliant on MKL details


> Once CI is happy, I think it makes sense to split off the changes to `.c|h|pyf` files and merge those, since that can immediately be done.

CI is now, and some the majority of `c|h|pyf` changes are in; 
but let's maybe do it the other way around and start from the build system changes.
This PR then is a minimal reasonable product (MRP): it has the needed tweaks to meson.build files + minimal source changes to have
scipy.linalg working.
Once we're happy with this (am ready to trim more fat from this one if you want me to),
I'll work on making ILP64 work in some of the newer C submodules. WDYT?

ev-br commented:
> Is it going to be compile-time or runtime switch you have in mind?

Build time only, sorry. Need to have MKL when building scipy. Can you build with MKL locally?

 > Not sure how to help but great that this is happening. The new C modules are pretty much in the "find/replace int -> ILP_INT" state. But the calls to BLAS/LAPACK, I don't know if possible to do if the symbol names are going to be varying.

The basic idea is in interpolate in this PR. Let's iron it out on one of the additions, shall we?
It'd be best to have an example which clearly fails if LAPACK is 32-bit and passes if it is  64-bit.

rgommers commented:
> This PR then is a minimal reasonable product (MRP): it has the needed tweaks to meson.build files + minimal source changes to have
> scipy.linalg working.

Ah okay, I am also fine with trying to merge this whole PR as long as we're a bit careful with how it's advertised. The CI job now only works because of:
```
pip install git+https://github.com/numpy/meson.git@main-numpymeson
```
which means the new build options need a disclaimer to not use/rely on them for distro packaging etc., they are subject to change.

ev-br commented:
> But the calls to BLAS/LAPACK, I don't know if possible to do if the symbol names are going to be varying.

Here's a branch: https://github.com/ev-br/scipy/compare/ilp64...ev-br:scipy:nnls_ilp64?expand=1
Currently it passes all tests locally with the "usual" lapack; `test_nnls.py` passes with ILP64, and SLSQP segfaults with it :-). I rather strongly suspect the culprit is in rather obscure `SLSQP_vars` handling, which involves `PyLong_FromLong` etc and those are quite unforgiving to C int type mismatches. Anyhow, these things is one reason why I think it makes sense to defer porting these C extensions to a separate PR.


> a bit careful with how it's advertised. The CI job now only works because of:
   >> pip install git+https://github.com/numpy/meson.git@main-numpymeson

Sure, it's ways off from being ready for distro packaging. For now it's just "there exists at least one not completely insane setup where it works". That said, the need for a numpy meson fork is orthogonal to scipy-internal code/build changes, I'd think. Once we have this single CI job, we can incrementally iron out smaller issues, generalize to other lapacks etc. And once numpy's meson fork lands upstream, we just stop using the fork and all is well.

ilayn commented:
That's a very common Python dict <-> C struct mapping. So if there is a problem it should error out but I think it might be that LongLong is not necessarily 64 bit. Separate PR also seems good to me so that we don't conflate with side effects.

Also if we are doing this, we should map to exact integer types from `stdint.h` https://en.cppreference.com/w/c/types/integer so that we never fall into these architectural issues.

rgommers commented on scipy/_build_utils/src/wrap_g77_abi_ilp64.c:
```diff
@@ -0,0 +1,114 @@
+ * The only difference between this and wrap_g77_abi.c is that here we
+ * export the wrappers mangled with BLAS_FUNC (i.e. with BLAS_PREFIX) instead
+ * of mangling with F_FUNC from fortran_defs.h.
+ * This mirrors the usage in scipy/linalg/fblas_64.pyf.src
```
Did you consider avoiding the duplication by, for example, defining a new `WRAP_BLAS_FUNC` which is an alias to either `BLAS_FUNC` or `F_FUNC` depending on whether `HAVE_BLAS_ILP64` is defined?

rgommers commented on scipy/interpolate/meson.build:
```diff
@@ -123,25 +128,33 @@ py3.extension_module('_rgi_cython',
     ['src/__fitpack.h', 'src/__fitpack.cc'],
-    dependencies:[lapack, np_dep, py3_dep],
+    dependencies:[lapack_ilp64, np_dep, py3_dep],
+    c_args: c_flags_ilp64
```
This looks wrong. It's possible that the code is actually correct, but only if `lapack_ilp64` can also contain an LP64 lapack. In which case it's still confusing; best to have a cleaner solution in `scipy/meson.build` then - both these variables need renaming at a minimum.

rgommers replied:
`c_flags_ilp64` can probably also be wrapped into the `lapack` object in `scipy/meson.build` so that there's no way to get it wrong anywhere. It's not good practice to carry around required flags that go with a dependency in a separate variable. Something like (untested):
```meson
lapack = declare_dependency(
    dependencies: lapack,
    compile_args: c_flags_ilp64,
)
```

rgommers replied:
Same comment applies to `_fflag_lp64`, `_fflag_ilp64`

ev-br replied:
> This looks wrong. It's possible that the code is actually correct, but only if lapack_ilp64 can also contain an LP64 lapack.

Almost. Does not contain it, but may be aliased to it
- `lapack_dep` is always LP64
- `lapack_ilp64` is ILP64 if available, equals `lapack_dep` otherwise (https://github.com/scipy/scipy/pull/22743/files#diff-b39b7bda5a39c101be8d9e0b3b6d4da335b6daf7385c2a43463125b0a1f33778R445)

Then a C extension links to `lapack_dep` if it's always LP64, or to `lapack_ilp64` if it is prepared to switch between LP64 and ILP64, if available.

Am open to a better naming scheme if this one's confusing.
----------------

An alternative is, of course, to have `lapack_ilp64` to be always ILP64 or not defined; then extensions will need to if-else in their `py3.extension_module` stanzas.

ev-br replied:
> c_flags_ilp64 can probably also be wrapped into the lapack object in scipy/meson.build so that there's no way to get it wrong anywhere

Done for this specific instance: things which depend on `lapack_ilp64` indeed get the right flags already (via https://github.com/scipy/scipy/pull/22743/files#diff-b39b7bda5a39c101be8d9e0b3b6d4da335b6daf7385c2a43463125b0a1f33778R442)

Other instances of `c_flags_ilp64` and `_fflag_ilp64`  I'm not sure how to remove: they are used to build fitpack_lib, which needs both C and Fortran build flags, and does not depend on `blas_ilp64`. Note that this is specific to a Fortran library which does not use BLAS/LAPACK.
Not sure what is the best practice is then. Make it artificially depend on `blas_ilp64` or make a special dependency just for the flags (how is this better than passing the flags around, as is currently done), or just leave as is. Happy to follow the best practice, just not sure what it is.

rgommers commented on scipy/interpolate/meson.build:
```diff
@@ -123,25 +128,33 @@ py3.extension_module('_rgi_cython',
+#      others rely on the build flags only. Consider cleaning this up at some point.
+
+
+_fitpack = py3.extension_module('_fitpack',
```
Defining `_fitpack` here doesn't seem to do anything. Please consider reverting to what it was before:
```meson
py3.extension_module('_fitpack',
```

rgommers commented on scipy/interpolate/meson.build:
```diff
@@ -151,12 +164,20 @@ py3.extension_module('_fitpack',
+   # generator only accepts strings, not files
+   f2c_map_file = f2py_ilp64_opts[1]
+   extra_arg = f2py_ilp64_opts[0] + '=' + fs.parent(f2c_map_file) / fs.name(f2c_map_file)
+  _dfitpackmodule_obj = f2py_gen.process('src/dfitpack.pyf', extra_args: extra_arg)
```
This deserves an explanation in a code comment for why only `dfitpack` needs this `f2c_map_file`.

rgommers commented on scipy/linalg/flapack_64.pyf.src:
```diff
@@ -16,9 +16,23 @@
 python module _flapack_64
     usercode '''
-#if defined(BLAS_SYMBOL_PREFIX) || defined(BLAS_SYMBOL_SUFFIX)
-#include "blas64-prefix-defines.h"
```
xref https://github.com/scipy/scipy/pull/19816#issuecomment-1879792156 for what this used to do

(no change to the code here needed)

rgommers replied:
That said, the lines below are now repeated twice, for `fblas_64` and `flapack_64` - would be good to put that into a new header.

rgommers commented on scipy/linalg/meson.build:
```diff
@@ -66,9 +66,12 @@ fblas_module = custom_target('fblas_module',
   dependencies: [lapack_dep, blas_dep, fortranobject_dep],
+  link_with: [g77_abi_wrappers],
   install: true,
+  link_language: 'fortran',
```
Question: why did this change? It seems wrong, this should be a pure C module.

ev-br replied:
No reason apparently. Is likely a leftover from past iterations. Removed now, thanks!

rgommers reviewed (commented):
Thanks Evgeni. Overall looks pretty reasonable. Here is a first set of review comments.

ev-br committed:
BLD: enable/test the ILP64 build with scipy-openblas64

ev-br committed:
MAINT: address review comments

ev-br committed:
MAINT: linalg: deduplicate the f2py defines

ev-br committed:
MAINT: deduplicate wrap_g77_abi{_ilp64}.c

ev-br committed:
DOC: document the status of the ILP64 support

ev-br committed:
MAINT: remove a duplicate build flag

ev-br committed:
MAINT: special: fix borked rebase

ev-br committed:
DOC: linalg: dev.py -> spin

ev-br commented:
Rebased on main to resolve conflicts this PR picked up, and to replace `dev.py` with `spin` in the docs.

ev-br committed:
CI: remove unused workflow

ev-br commented on meson.options:
```diff
@@ -2,6 +2,12 @@ option('blas', type: 'string', value: 'openblas',
+option('blas-symbol-suffix', type: 'string', value: 'auto',
+        description: 'BLAS and LAPACK symbol suffix to use, if any')
+option('mkl-threading', type: 'string', value: 'auto',
+        description: 'MKL threading method, one of: `seq`, `iomp`, `gomp`, `tbb`')
```
Done in a code comment just above. Or do want to have this wording in the `description` field?

ev-br committed:
DOC: linalg: use a correct `spin` incantation for -Doption

ev-br committed:
CI: use spin not dev.py

rgommers commented on .github/workflows/linux_blas_ilp64.yml:
```diff
@@ -0,0 +1,129 @@
+        pip install -r requirements/dev.txt
+        pip install git+https://github.com/numpy/meson.git@main-numpymeson
+        pip install scipy-openblas32 scipy-openblas64
+        # spin does this for scipy-openblas32
```
Can you add a separate job step for this?
```
name: Write out scipy-openblas64.pc
```

rgommers commented on .github/workflows/linux_blas_ilp64.yml:
```diff
@@ -0,0 +1,129 @@
+        pip install scipy-openblas32 scipy-openblas64
+        # spin does this for scipy-openblas32
+        python -c'import scipy_openblas64 as so64; print(so64.get_pkg_config())' > scipy-openblas64.pc
+        export PKG_CONFIG_PATH=`pwd`
```
This `export` should not be doing anything, since it doesn't get persisted between job steps. You normally need `echo xxx >> "$GITHUB_ENV"`. If things work for this job, you should be able to delete this line.

rgommers commented on doc/source/building/blas_lapack.rst:
```diff
@@ -96,6 +96,44 @@ user wants to override this autodetection mechanism for building against plain
+----------------------------------
+
+Support for ILP64 BLAS and LAPACK is still experimental; at the time of writing
+(Apr 2025) it is only available for two BLAS/LAPACK configurations: MKL and
```
can be updated to (Jul 2025)

rgommers reviewed (commented):
If MKL works, I'd expect Accelerate to work as well pretty much unchanged. Did you test? If not, should I?

rgommers reviewed (changes requested):
Looks pretty close - I have lots of comments, but they should be easy to address.

ev-br commented on scipy/_build_utils/src/wrap_g77_abi.c:
```diff
@@ -25,6 +25,12 @@ return values, struct complex arguments work without segfaulting.
 
+#ifdef HAVE_BLAS_ILP64
+/* NB: this redefines F_FUNC */
+#include "_blas64_defines.h"
```
I don't see any compiler warnings locally on linux at least. Not sure what the problem is that you're after. We need an `#include "npy_cblas.h"` for both LP64 and ILP64 builds, so in principle we can remove one include from `_blas64_defines.h` but why.

rgommers replied:
I was thinking `F_FUNC` was in `npy_cblas.h`, but it isn't. That said, this still looks very odd - I'm unclear on the logic. `F_FUNC` goes from _only_ dealing with Fortran uppercase/append to dealing with BLAS prefix/suffix, which isn't its original function at all.

Its original purpose is, from `_build_utils/src/fortran_defs.h`:
```c
/*
 * Handle different Fortran conventions.
 */

#if defined(NO_APPEND_FORTRAN)
#if defined(UPPERCASE_FORTRAN)
#define F_FUNC(f,F) F
#else
#define F_FUNC(f,F) f
#endif
#else
#if defined(UPPERCASE_FORTRAN)
#define F_FUNC(f,F) F##_
#else
#define F_FUNC(f,F) f##_
#endif
#endif
```

After you redefine it in `_blas64_defines.h` it becomes:

```c
#define BLAS_FUNC(name) BLAS_FUNC_EXPAND(name,BLAS_SYMBOL_PREFIX,BLAS_SYMBOL_SUFFIX,BLAS_FORTRAN_SUFFIX)
#define F_FUNC(f, F) BLAS_FUNC(f)
```

I'm not following that logic.

ev-br replied:
There's a brief discussion in the commit message of https://github.com/scipy/scipy/pull/22743/commits/76b30881a0d2d1030bb2b369ccaf599fa581025b, but I see it's too brief, so let me try rephrasing.

In the f2py-generated code, `F_FUNC` macro comes from f2py itself: https://github.com/numpy/numpy/blob/main/numpy/f2py/cfuncs.py#L112
The name of the macro and its content are fixed but f2py itself and I don't see a way to change it. Now, if you look into the f2py-generated `build/scipy/linalg/_fblas64_module.c`:

```
/********************** See f2py2e/cfuncs.py: cppmacros **********************/

#if defined(PREPEND_FORTRAN)
#if defined(NO_APPEND_FORTRAN)
#if defined(UPPERCASE_FORTRAN)
#define F_FUNC(f,F) _##F
#else
#define F_FUNC(f,F) _##f
#endif
#else
#if defined(UPPERCASE_FORTRAN)
#define F_FUNC(f,F) _##F##_
#else
#define F_FUNC(f,F) _##f##_
#endif
#endif
#else
#if defined(NO_APPEND_FORTRAN)
#if defined(UPPERCASE_FORTRAN)
#define F_FUNC(f,F) F
#else
#define F_FUNC(f,F) f
#endif
#else
#if defined(UPPERCASE_FORTRAN)
#define F_FUNC(f,F) F##_
#else
#define F_FUNC(f,F) f##_
#endif
#endif
#endif

// ... snip ...

/********************* See f2py2e/capi_rules.py: usercode *********************/
  /* start usercode multiline (0) */
 
#include "_blas64_defines.h"

  /* end multiline (0)*/

/* See f2py2e/rules.py */
extern void F_FUNC(srotg,SROTG)(float*,float*,float*,float* );
extern void F_FUNC(drotg,DROTG)(double*,double*,double*,double* );
extern void F_FUNC(crotg,CROTG)(complex_float*,complex_float*,complex_float*,complex_float* );
extern void F_FUNC(zrotg,ZROTG)(complex_double*,complex_double*,complex_double*,complex_double* );

//....
```

So what happens is that f2py defines the `F_FUNC` macro (we cannot change it), then we include `_blas64_defines.h` from the usercode section, and then `F_FUNC` macro is used to mangle the BLAS names.

So in the usercode section, we undefine the not-useful `F_FUNC` macro and redefine it with `BLAS_FUNC` from `npy_cblas.h` --- which does the LP64/ILP64 name mangling we want.

rgommers replied:
Ah thanks - that would be really helpful to put in a code comment.

ev-br replied:
Any suggestions as to where to add this comment? (I stared for too long too recently into this, my eyes are biased)

rgommers replied:
Maybe extend the code comment in `_blas_defines.h` with what's in the commit message?

ev-br replied:
Thanks, done.

rgommers replied:
Great, resolving this

ev-br commented on scipy/meson.build:
```diff
@@ -335,6 +369,104 @@ else
+  blas_ilp64 = blas
+  lapack_ilp64 = lapack
+  c_flags_ilp64 = []
+  message('LAPACK ILP64 not requested / not detected.')
```
Confusing yes, helpful also yes though.

The key issue is we need to somehow express "link to 64-bit LAPACK if available else to 32-bit LAPACK". Therefore there's a need for an `if-else` somewhere. Currently it's here, so that all extensions simply depend on `lapack_ilp64` which may be ILP64 or LP64 depending on the meson option.

Alternatives:
- rename the dependency to `lapack_maybe_ilp64` or some such, or
- push the if-else to individual extensions  (`dependcies: [lapack_ilp64 if available lapack_dep otherwise]`).

Are you suggesting the second variant?

rgommers replied:
No, I'm not suggesting the second variant. We need clear names for the first variant. We have `2**3` options:
- BLAS and LAPACK
- LP64-only and maybe-ILP64
- C/C++ and C/C++/Fortran

So how about we have eight `_dep` objects, and each build target will depend on exactly one of those objects with no other dependencies or compile arg objects needing to be supplied?

Names can be: 
- `blas_lp64_dep`
- `blas_dep`
- `blas_lp64_fortran_dep`
- `blas_fortran_dep`

and the same for `lapack_xxx`? That way `blas_dep` is the generic one, may be LP64 or ILP64 which everything should support in the future; `blas_lp64` is the special case. And the `_fortran` flavors will be disappearing completely over the next year.

Also `lapack_xxx_dep` should include `dependencies: [blas_xxx]` so that we no longer have to depend on both blas and lapack anywhere.

rgommers replied:
Okay great, let's drop the Fortran variants - let's just add a code comment above the four `_dep` objects that we didn't create those due to limited need, but it could be done if Fortran usage unexpectedly proliferates.

> So it is then basically a rename, `lapack_dep -> lapack_lp64_dep` and `lapack_ilp64 -> lapack_dep`. WDYT?

SGTM

rgommers replied:
Thanks, this looks good now - resolving.

ev-br reviewed (commented):
Thanks for the review @rgommers. Let me clarify two larger items first, (and I'll address smaller items in a separate update)

ev-br committed:
MAINT: ILP64: address review comments

ev-br committed:
MAINT: f2py: use #define not typedef

ev-br committed:
MAINT: interpolate: simplify the f2py 64-bit arg handling

ev-br committed:
MAINT: linalg: remove an extra explicit link_with: g77_abi_wrappers

ev-br committed:
MAINT: purge last vestiges of "blas" dependency (always use blas_dep instead)

ev-br committed:
MAINT: add missing - to -cpp

ev-br committed:
WIP: _fflag_ilp64: not quite right still (arnaud gets extra Fortran args)

ev-br committed:
DOC: add a link to an upstream report for MKL

ev-br committed:
BUG: sparse/linalg: arnaud is not ILP64-ready

ev-br commented on scipy/meson.build:
```diff
@@ -504,13 +636,51 @@ fortran_ignore_warnings = ff.get_supported_arguments(
+  # that depend on BLAS/LAPACK
+  # Note: this came from `scipy/build_utils/_fortran.py` in the distutils build.
+  _fflag_lp64 += _fflag_preprocess  # TODO: propagate _fflag_lp64
+  _fflag_ilp64 += _fflag_preprocess
```
I tried this in 814dc4634dae82375eb216eaa008bb57120d472d but the compile flag gets added to both Fortran and C sources, which then triggers

```
[15/19] Compiling C object scipy/linalg/_fblas_64.cpython-312-x86_64-linux-gnu.so.p/meson-generated_..__fblas_64module.c.o
cc1: warning: command-line option '-fdefault-integer-8' is valid for Fortran but not for C
```

What would be a way only apply `compile_args` to Fortran sources and not C sources?

rgommers replied:
Oh grumble, yeah I've run into that before, there's no language-specific `compile_args`. We'll need two separate `_dep` objects, one for C/C++ and one for targets that also contain Fortran code.

Let me actually test the latest changes and then circle back to this.

ev-br replied:
> We'll need two separate _dep objects, one for C/C++ and one for targets that also contain Fortran code.

To make it more "fun", some targets need `-fdefault-integer-8` but do not link to LAPACK (dfitpack is a prime example). 
All in all, if we expect that all fortran targets are gone in a not-so-distant future, I can't help wondering if the cure is not worse than the decease

rgommers replied:
> To make it more "fun", some targets need `-fdefault-integer-8` but do not link to LAPACK (dfitpack is a prime example).

`dfitpack` does link to lapack?

ev-br replied:
No, it doesn't

ev-br replied:
It was on main, but it's not needed and is removed in this PR: 
https://github.com/scipy/scipy/pull/22743/files#diff-85ab739d8d3eb989d275834cd1da1d3302bedb33b8ecbef8656831d6ad026edbR176
(no idea why it was there at all, might have been my previous error)

ev-br committed:
DOC: linalg: expand the ILP64 section of the docs

rgommers commented on .github/workflows/linux_blas_ilp64.yml:
```diff
@@ -0,0 +1,129 @@
+name: BLAS tests (Linux)
```
These jobs aren't running: https://github.com/scipy/scipy/actions/workflows/linux_mkl_ilp64.yml

ev-br committed:
CI: fix the CI ILP64 job

ev-br committed:
MAINT/BLD: combine blas_dep and lapack_dep into a single dependency object

ev-br committed:
DOC/BLD: extend a comment in _blas64_defines.h

ev-br committed:
MAINT: rename lapack_dep -> lapack_lp64_dep

ev-br committed:
MAINT: rename lapack_ilp64 -> lapack_dep

From now on, lapack_dep is ILP64 if available, LP64 otherwise.
An extension should depend on
- `lapack_lp64_dep` if it is LP64-only
- `lapack_dep` if it is ready to build-time switch between LP64 and ILP64 variants

ev-br committed:
MAINT: clean up the usage of fflags_ilp64, c_flags_ilp64

A C extension now only needs to add a dependency on [lapack_dep] to be
LP64 / ILP64 ready

rgommers commented on scipy/_build_utils/meson.build:
```diff
@@ -12,7 +12,7 @@ if generate_blas_wrappers
     blas_lapack_wrappers,
     c_args: numpy_nodepr_api,
-    dependencies: [blas_dep, lapack_dep, np_dep, py3_dep],
+    dependencies: [lapack_lp64_dep, np_dep, py3_dep],
```
This can't be right, can it? The purpose of this wrapper should be at least as relevant for ILP64. xref gh-19816 for why this was introduced, it already mentions ILP64.

ev-br replied:
It is certainly not needed for MKL and OpenBLAS, and I cannot easily test Accelerate. Maybe it's the missing bit for Accelerate; however `generate_blas_wrappers = true` for openblas, too, and things work as they are now. Not sure.

ev-br replied:
Actually, 

```
$ cat build/scipy/_build_utils/blas_lapack_wrappers.c  |more
/*
This file was generated by _generate_blas_wrapper.py.
Do not edit this file directly.
*/


#include "npy_cblas.h"
#include "fortran_defs.h"

// ... snip ....

void BLAS_FUNC(caxpy)(int *n, npy_complex64 *ca, npy_complex64 *cx, int *incx, npy_complex64 *cy, int *incy);
void F_FUNC(caxpy,CAXPY)(int *n, npy_complex64 *ca, npy_complex64 *cx, int *incx, npy_complex64 *cy, int *incy){
    return BLAS_FUNC(caxpy)(n, ca, cx, incx, cy, incy);
}

// ...

```

So this basically makes a bunch of trampolines from `F_FUNC` to `BLAS_FUNC`? If so, the ILP64 version is using the `#undef F_FUNC` construction instead which we discussed here yesterday.

rgommers replied:
> So this basically makes a bunch of trampolines from `F_FUNC` to `BLAS_FUNC`? If so, the ILP64 version is using the `#undef F_FUNC` construction instead which we discussed here yesterday.

I think that's part of why that confused me. All symbol prefixes/suffixes were handled by this generated wrapper. ILP64 adds symbol suffixes, so I expected it to be handled by these wrappers as well (of course that's made more complex by us needing both LP64 and ILP64). Don't we now have two similar but different ways of doing the same thing?

In addition, `_generate_blas_wrappers.py` it generates wrappers for all LAPACK functions as well. So I think what could be done in principle is to call `_generate_blas_wrappers.py` twice, once for LP64 wrappers and once for ILP64 wrappers - after updating `scipy/_build_utils/_wrappers_common.py` and whatever that pulls in for ILP64 (so we'd have `cython_lapack64_signatures.txt` in addition to `cython_lapack_signatures.txt`).

- `_blas64_defines.h` is reintroducing what we had with `distutils` way back when, but is specific to `fblas_64.pyf.src` and `flapack_64.pyf.src`, and our `f2py` signature file method.
- The `blas_lapack_wrappers` method is using generated wrappers for the Cython API, **and** was used for all symbol suffix/prefix handling in current LP64 usage (`blas_dep`, `lapack_dep`). 
- xref gh-20682 which is related to the difference between these methods.
 
* * *

Okay, I think I finally have it sharp again how this all worked. I think we should merge this PR as is, but also discuss this topic further since we now both are looking at it in detail. This brings a bunch of extra tech debt as it stands (disclaimer: in my current understanding, which may still be missing something) because it splits suffix/prefix handling into two and leaves ILP64 LAPACK incomplete while LP64 LAPACK is complete.

ev-br committed:
MAINT: remove fflag_lp64 (is always empty)

rgommers commented:
Okay, CI is happy and review comments were addressed, so I'll do one last pass with maybe a few last cleanups, and look into Accelerate locally, then get this in.

@ev-br the commit history is a bit messy, I'm considering squash-merging it - unless you have a different preference?

ev-br commented:
> @ev-br the commit history is a bit messy, I'm considering squash-merging it - unless you have a different preference?

+1 to squash merge, yes.

ilayn commented:
Regarding the Cython based BLAS/LAPACK we learned to mimic f2py by stride tricks and silent F-major conversions etc. in our linalg work. Hence we can follow the

argument sanitizing -> Cython wrapper -> BLAS/LAPACK linkage

instead of 

f2py -> argument sanitizing -> Fortran wrapper -> BLAS LAPACK linkage

That's basically the nice thing to remove Fortran compiler dependence in its entirety contingent upon the remaining code disappearing.

rgommers commented:
I did a full set of wheel builds [CI log](https://github.com/rgommers/scipy/actions/runs/16352000484) and local testing for Accelerate - all looks fine. 

A few more little things to fix up, I'll push them to this branch.

Could you open a new tracking issue for per-module ILP64 support wherever that needs to be added back?

rgommers committed:
BUG: fix `scipy.show_config` output

Also some other minor cleanups.

rgommers committed:
MAINT: rename `linux_blas_ilp64.yml` to `linux_blas.yml`

There is already an LP64 build in there, and we'll use this workflow
file for general BLAS testing.

rgommers reviewed (approved):
Submitting last few review comments for which I just pushed two commits.

ev-br commented:
> Could you open a new tracking issue for per-module ILP64 support wherever that needs to be added back?

https://github.com/scipy/scipy/issues/23351
